PHP 8.3 support added

glo71317 opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes/no
| Bugfix        | yes/no
| BC Break      | yes/no
| New Feature   | yes/no
| RFC           | yes/no
| QA            | yes/no

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding documentation?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

glo71317 commented:
Hi @Ocramius 

Few tests are failing due to some certification issue
**Error -** [Microsoft][ODBC Driver 18 for SQL Server]SSL Provider: [error:0A000086:SSL routines::certificate verify failed:self-signed certificate]

It seems some issue with CI/CD Any plan to fix this issue OR anything can we do?

Ocramius commented:
Yep, the usual MSSQL mess.

@weierophinney can we perhaps drop testing MSSQL? It's such a massive waste of time and oxygen for other better RDBMS...

glo17680 commented:
@Ocramius / @weierophinney  is there any update on this. Can anyone please advise on the same ?

Ocramius commented:
Either we rip out MSSQL testing (and mark it as untested somewhere in the docs), or we wait for Microsoft to release MSSQL support for PHP 8.3, which usually happens few weeks/months after release.

weierophinney commented:
My feelings on the matter are:

- Not testing against MS SQL Server means we don't support it
- Not supporting it would be a BC break, requiring a new major
- We've marked this component as security-only, so a new major feels like a break with that

As such, my vote is we wait until MS has released a version of the extension for 8.3. If tests still fail at that time, we can regroup and figure out what to do from there; my guess is everything will "just work" as it has the last several releases.

glo71317 committed:
Adds Support New PHP 8.3

Signed-off-by: Rajesh Kumar <glo71317@adobe.com>

glo71317 commented:
@Ocramius @weierophinney Any update on this?

Ocramius commented:
No.

driehle commented:
This is a bit unfortunate, as this blocks [laminas-log](https://github.com/laminas/laminas-log/pull/53), which is quite a central component in many applications.

How about extracting the namespace `Laminas\Db\Adapter\Driver\Sqlsrv` into a separate package? This would, obviously, require a new major release.

weierophinney commented:
@driehle Considering this repo is security-only, we're not going to do a new major. It would likely make more sense to split the lamians-db related functionality in laminas-log into a new package.

I recognize folks are excited about the new PHP release but (a) it's been not even two weeks, and (b) not every organization is tracking PHP updates in order to release as soon as it drops. It's far better for us to wait and ensure that the component works as expected (which means running tests!) than it is to rush a release. Please be patient.

glo71317 commented:
@Ocramius Any update on this when we can expect laminas-db available with php 8.3 support

Ocramius commented:
https://github.com/laminas/laminas-db/pull/293#issuecomment-1780945285

rchrdkn commented:
Looks like this is the msphpsql issue that we're waiting on - https://github.com/microsoft/msphpsql/issues/1477

Let's hope they decide to release soon 🤞🏻

panvid commented:
There is noting new?

glo71317 commented:
@Ocramius @weierophinney https://github.com/microsoft/msphpsql/releases/tag/v5.12.0 this released Can we prioritize this?

glo71317 commented:
I have fixed all the test failures and other deprecated dynamic properties issues, Only following error is there 

> Unable to connect laminasdb_test. Array
```
(
    [0] => Array
        (
            [0] => 08001
            [SQLSTATE] => 08001
            [1] => -1
            [code] => -1
            [2] => [Microsoft][ODBC Driver 18 for SQL Server]SSL Provider: [error:0A000086:SSL routines::certificate verify failed:self-signed certificate]
            [message] => [Microsoft][ODBC Driver 18 for SQL Server]SSL Provider: [error:0A000086:SSL routines::certificate verify failed:self-signed certificate]
        )
    [1] => Array
        (
            [0] => 08001
            [SQLSTATE] => 08001
            [1] => -1
            [code] => -1
            [2] => [Microsoft][ODBC Driver 18 for SQL Server]Client unable to establish connection. For solutions related to encryption errors, see https://go.microsoft.com/fwlink/?linkid=2226722
            [message] => [Microsoft][ODBC Driver 18 for SQL Server]Client unable to establish connection. For solutions related to encryption errors, see https://go.microsoft.com/fwlink/?linkid=2226722
        )
)
```

it seems it should be fixed from your end or suggest to us we can do anything for this.

Ocramius commented:
SSounds like a new requirement by the MSSQL adapter? Can you try disabling SSL validation for MSSQL in the test suite?

Xerkus commented:
This has to be done via connection options https://learn.microsoft.com/en-us/sql/connect/php/connection-options by setting `TrustServerCertificate` to 1

This is introduced by ODBC Driver 18

glo71317 commented:
> SSounds like a new requirement by the MSSQL adapter? Can you try disabling SSL validation for MSSQL in the test suite?

No idea how to disable ssl validation for MSSQL in the suite. so any idea?

driehle commented:
https://github.com/laminas/laminas-db/blob/12385a47561c2d1feadf2bd114a67767541963a3/test/integration/Adapter/Platform/SqlServerTest.php#L36-L40

As Xerkus said, try adding `'TrustServerCertificate' => 1` there.

glo71317 commented:
> https://github.com/laminas/laminas-db/blob/12385a47561c2d1feadf2bd114a67767541963a3/test/integration/Adapter/Platform/SqlServerTest.php#L36-L40
> 
> As Xerkus said, try adding `'TrustServerCertificate' => 1` there.

I have made changes - https://github.com/laminas/laminas-db/pull/293/commits/2abd29d160bdca714e8198e786d4a1dcfcd0c4a4 but still having same issue

Xerkus commented:
There is more than one place where `sqlsrv_connect()` is used. You need to find which of those need the option

Xerkus commented on .laminas-ci/install-sqlsrv-extension-via-pecl.sh:
Please drop this file entirely. sqlsrv is provided by CI container.

glo71317 replied:
done

glo71317 commented:
> There is more than one place where `sqlsrv_connect()` is used. You need to find which of those need the option

I have updated all the places and fixed that issues, now lot tests are failing, Can you check these need to fixed still having some issue?

Xerkus commented on .laminas-ci.json:
```diff
@@ -9,6 +9,6 @@
         "sqlsrv"
     ],
     "ignore_php_platform_requirements": {
-        "8.1": true
+        "8.3": true
     }
```
Please remove platform ignores, whole thing

Xerkus commented:
Errors are caused by phpunit deprecations. Tests need to be migrated to newer phpunit version.

Xerkus commented:
There is still one certificate validation error: 
```
 1) LaminasIntegrationTest\Db\Adapter\Platform\SqlServerTest::testQuoteValueWithSqlServer
PDOException: SQLSTATE[08001]: [Microsoft][ODBC Driver 18 for SQL Server]SSL Provider: [error:0A000086:SSL routines::certificate verify failed:self-signed certificate]

/github/workspace/test/integration/Adapter/Platform/SqlServerTest.php:54
```

Xerkus commented:
> I have updated all the places and fixed that issues, now lot tests are failing, Can you check these need to fixed still having some issue?

Those are phpunit deprecation warnings. Tests need to be migrated to latest phpunit but i do not think they cause CI to fail. They would need to be dealt with in a separate PR.

glo71317 commented:
@Ocramius Can you review this PR changes and @Xerkus comments? what is your point of view on this?

Xerkus commented on src/Adapter/Driver/Sqlsrv/Connection.php:
```diff
@@ -101,6 +101,7 @@ public function connect()
         $serverName = '.';
         $params     = [
             'ReturnDatesAsStrings' => true,
+            'TrustServerCertificate' => 1
```
This is a security feature that can not be applied as a default for adapter. Please revert.
This connection parameter needs to be properly configured via `'options'` from test case.

glo71317 replied:
reverted

Xerkus commented on .laminas-ci.json:
```diff
@@ -9,6 +9,6 @@
         "sqlsrv"
     ],
     "ignore_php_platform_requirements": {
-        "8.1": true
+        "8.3": true
     }
```
Remove platform ignores. There were needed for php release candidate and should not be present now.

glo71317 replied:
done

Xerkus commented on .laminas-ci/install-sqlsrv-extension-via-pecl.sh:
Please remove this script. It is not needed.

sqlsrv extension is available in CI container directly. This script compiles extension from pecl and significantly increases CI run time.

glo71317 replied:
done

Xerkus commented on phpcs.xml.dist:
```diff
@@ -32,4 +32,38 @@
         <exclude-pattern>src/Adapter/AdapterServiceFactory.php</exclude-pattern>
         <exclude-pattern>src/Adapter/AdapterAbstractServiceFactory.php</exclude-pattern>
     </rule>
+    <rule ref="WebimpressCodingStandard.Arrays.DoubleArrow.SpacesBefore">
+        <exclude-pattern>test/integration/Platform/SqlServerFixtureLoader.php</exclude-pattern>
+        <exclude-pattern>test/integration/Adapter/Platform/SqlServerTest.php</exclude-pattern>
+        <exclude-pattern>test/unit/Adapter/Driver/Sqlsrv/ConnectionIntegrationTest.php</exclude-pattern>
+        <exclude-pattern>src/Adapter/Driver/Sqlsrv/Connection.php</exclude-pattern>
+        <exclude-pattern>test/unit/Adapter/Driver/Sqlsrv/AbstractIntegrationTest.php</exclude-pattern>
+        <exclude-pattern>test/unit/Adapter/Driver/Sqlsrv/StatementIntegrationTest.php</exclude-pattern>
+    </rule>
+    <rule ref="SlevomatCodingStandard.ControlStructures.RequireNullCoalesceOperator.NullCoalesceOperatorNotUsed">
+        <exclude-pattern>src/Adapter/Platform/Sql92.php</exclude-pattern>
+    </rule>
+    <rule ref="WebimpressCodingStandard.Arrays.TrailingArrayComma.MissingTrailingComma">
+        <exclude-pattern>test/unit/Adapter/Driver/Sqlsrv/ConnectionIntegrationTest.php</exclude-pattern>
+        <exclude-pattern>test/integration/Platform/SqlServerFixtureLoader.php</exclude-pattern>
+        <exclude-pattern>test/integration/Adapter/Platform/SqlServerTest.php</exclude-pattern>
+        <exclude-pattern>src/Adapter/Driver/Sqlsrv/Connection.php</exclude-pattern>
+        <exclude-pattern>test/unit/Adapter/Driver/Sqlsrv/AbstractIntegrationTest.php</exclude-pattern>
+        <exclude-pattern>test/unit/Adapter/Driver/Sqlsrv/StatementIntegrationTest.php</exclude-pattern>
+    </rule>
+    <rule ref="SlevomatCodingStandard.Namespaces.ReferenceUsedNamesOnly.ReferenceViaFullyQualifiedName">
+        <exclude-pattern>test/unit/RowGateway/RowGatewayTest.php</exclude-pattern>
+        <exclude-pattern>test/unit/Sql/Platform/Mysql/SelectDecoratorTest.php</exclude-pattern>
+        <exclude-pattern>test/unit/Adapter/Driver/IbmDb2/StatementTest.php</exclude-pattern>
+        <exclude-pattern>test/unit/TableGateway/AbstractTableGatewayTest.php</exclude-pattern>
+    </rule>
+    <rule ref="Squiz.PHP.NonExecutableCode.ReturnNotRequired">
+        <exclude-pattern>src/Adapter/Driver/IbmDb2/Connection.php</exclude-pattern>
+        <exclude-pattern>src/Adapter/Driver/Oci8/Connection.php</exclude-pattern>
+        <exclude-pattern>src/Adapter/Driver/Mysqli/Connection.php</exclude-pattern>
+        <exclude-pattern>src/Adapter/Driver/Pgsql/Connection.php</exclude-pattern>
+    </rule>
+    <rule ref="Generic.WhiteSpace.IncrementDecrementSpacing.SpaceAfterIncrement">
+        <exclude-pattern>src/Adapter/Driver/Pdo/Connection.php</exclude-pattern>
+    </rule>
```
Coding style violations need to be fixed, not ignored. Please revert.

glo71317 replied:
reverted

Xerkus commented on test/integration/Adapter/Driver/Pdo/Postgresql/AdapterTest.php:
```diff
@@ -4,6 +4,7 @@
 
 use LaminasIntegrationTest\Db\Adapter\Driver\Pdo\AbstractAdapterTest;
 
+#[\AllowDynamicProperties]
```
I would prefer you to explicitly declare dynamically used properties rather than use this attriburte.
This annotation is hiding code quality issue and it is not suitable for this use case.

Same for all other uses of this attribute below.

Xerkus reviewed (changes requested):
Maintenance work is frustrating but necessary. Thank you for your effort.

Xerkus commented:
@glo71317 remaining error is because trust option is missing for pdo dsn:

```diff
--- a/test/integration/Adapter/Platform/SqlServerTest.php
+++ b/test/integration/Adapter/Platform/SqlServerTest.php
@@ -49,7 +49,7 @@ class SqlServerTest extends TestCase
             $this->adapters['pdo_sqlsrv'] = new PDO(
                 'sqlsrv:Server='
                     . getenv('TESTS_LAMINAS_DB_ADAPTER_DRIVER_SQLSRV_HOSTNAME')
-                    . ';Database=' . ($database ?: ''),
+                    . ';Database=' . ($database ?: '') .';TrustServerCertificate=1',
                 getenv('TESTS_LAMINAS_DB_ADAPTER_DRIVER_SQLSRV_USERNAME'),
                 getenv('TESTS_LAMINAS_DB_ADAPTER_DRIVER_SQLSRV_PASSWORD')
             );
```

glo71317 committed:
PHP 8.3 Support

Signed-off-by: Rajesh Kumar <glo71317@adobe.com>

glo71317 committed:
PHP 8.3 Support

Signed-off-by: Rajesh Kumar <glo71317@adobe.com>

Xerkus commented on src/Adapter/Driver/IbmDb2/Connection.php:
```diff
@@ -114,7 +114,7 @@ public function connect()
             }
 
-            return;
+            return; // phpcs:ignore
```
Again. Do not ignore coding standard issues. Fix them.

Xerkus replied:
Since return value is expected
 
```suggestion
            return null;
```

glo71317 replied:
Are you want to remove empty retrun from here? error showing `Empty return statement not required here`

glo71317 replied:
done

Xerkus replied:
There is a distinction between `return;` and `return null` or omitting return entirely.

When return types are declared both of those will result in a fatal error:

```
function (): void {
    return null;
};
```

```
function (): ?string {
    return;
};
```

In this case return type is not explicitly defined so php does not know which way is correct so there is no error.

We could technically omit return entirely for implicit return at the end.
When return value is not expected omitting `return;` is correct.
However, when return is expected but no return statement present we get a code quality issue since it is not entirely clear whether `return null` was implied or return statement was simply forgotten. Psalm normally would complain about this case.

glo71317 committed:
Fixed README Linting

Signed-off-by: Rajesh Kumar <glo71317@adobe.com>

glo71317 committed:
Fixed README Linting

Signed-off-by: Rajesh Kumar <glo71317@adobe.com>

glo71317 commented:
@Xerkus it seems tests are not running due to .laminas-ci/install-sqlsrv-extension-via-pecl.sh as you told to removed

Xerkus commented:
> @Xerkus it seems tests are not running due to .laminas-ci/install-sqlsrv-extension-via-pecl.sh as you told to removed

Yes, it is referenced here: https://github.com/laminas/laminas-db/blob/12385a47561c2d1feadf2bd114a67767541963a3/.laminas-ci/pre-install.sh#L19

Please also remove that line.

glo71317 committed:
PHP 8.3 support added

Signed-off-by: Rajesh Kumar <glo71317@adobe.com>

Xerkus commented on README.md:
```diff
@@ -3,19 +3,15 @@
 
 > ## 🇷🇺 Русским гражданам
-> 
+
```
This is not a correct. markdown lint complained about trailing whitespaces in `> `

Xerkus commented:
I asked you to revert changes in adapter previously.  The correct way is to pass connection options as parameters for  `Connection` instance: `new Connection([..., 'options' => ['TrustServerCertificate' => '1']]);`

I looked at the tests code and best place seems to be here: 
```diff
--- a/test/unit/Adapter/Driver/Sqlsrv/AbstractIntegrationTest.php
+++ b/test/unit/Adapter/Driver/Sqlsrv/AbstractIntegrationTest.php
@@ -38,6 +38,8 @@ abstract class AbstractIntegrationTest extends TestCase
             $this->variables[$name] = getenv($value);
         }
 
+        $this->variables['options'] = ['TrustServerCertificate' => '1'];
+
         if (! extension_loaded('sqlsrv')) {
             $this->fail('The phpunit group integration-sqlsrv was enabled, but the extension is not loaded.');
         }
```

Xerkus commented:
Also you seem to have missed re-adding changes for some ` sqlsrv_connect()` calls in tests.

glo71317 commented:
> sqlsrv_connect

I don't think as i can see here i removed `src/Adapter/Driver/Sqlsrv/Connection.php` as u suggested

Xerkus commented:
> > sqlsrv_connect
> 
> I don't think as i can see here i removed `src/Adapter/Driver/Sqlsrv/Connection.php` as u suggested

Nevermind, looking at a wrong commit.

glo71317 committed:
Fixed Deprecated: Creation of dynamic property

Signed-off-by: Rajesh Kumar <glo71317@adobe.com>

Xerkus commented on test/integration/Adapter/Platform/SqlServerTest.php:
```diff
@@ -48,7 +49,7 @@ protected function setUp(): void
                 'sqlsrv:Server='
                     . getenv('TESTS_LAMINAS_DB_ADAPTER_DRIVER_SQLSRV_HOSTNAME')
-                    . ';Database=' . ($database ?: ''),
+                    . '; Database=' . ($database ?: '') . '; TrustServerCertificate=1',
```
AFAIK spaces in DSN are invalid.

glo71317 replied:
If i removed than phpcs failed

glo71317 committed:
PHP8.3 support added

Signed-off-by: Rajesh Kumar <glo71317@adobe.com>

glo71317 commented:
All issues resolved from my end now  and pr is green now thank you so much for your quick support and review

Xerkus commented:
You are missing signoff for your commits. See DCO check for details

glo71317 commented:
> You are missing signoff for your commits. See DCO check for details

Shall i use following command
`git rebase HEAD~8 --signoff`
`git push --force-with-lease origin php8.3_support `

Xerkus commented on test/integration/Adapter/Platform/SqlServerTest.php:
```diff
@@ -48,7 +49,7 @@ protected function setUp(): void
                 'sqlsrv:Server='
                     . getenv('TESTS_LAMINAS_DB_ADAPTER_DRIVER_SQLSRV_HOSTNAME')
-                    . ';Database=' . ($database ?: ''),
+                    . '; Database=' . ($database ?: '') . '; TrustServerCertificate=1',
```
```suggestion
                    . ';Database=' . ($database ?: '') . ';TrustServerCertificate=1',
```

glo71317 replied:
done

glo71317 committed:
PHP8.3 support added

Signed-off-by: Rajesh Kumar <glo71317@adobe.com>

Xerkus committed:
Fix README formatting and update badge

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>
